src/Xamarin.Android.Build.Tasks/Tasks/JavaDependencyVerification.cs

```diff
@@ -0,0 +1,691 @@
+	{
+		item.TryParseJavaArtifactAndJavaVersion (itemName, logger, out var artifact, out var _);
+
+		if (!File.Exists (filename)) {
+			logger.LogCodedError ("XA4245", Properties.Resources.XA4245, filename);
+			return null;
+		}
+
+		try {
+			using (var file = File.OpenRead (filename)) {
+				var project = Project.Parse (file);
+				var registered_artifact = Artifact.Parse (project.ToString ());
+
+				// Return the registered artifact, preferring any overrides specified in the task item
+				var final_artifact = new Artifact (
+					artifact?.GroupId ?? registered_artifact.GroupId,
+					artifact?.Id ?? registered_artifact.Id,
+					artifact?.Version ?? registered_artifact.Version
+				);
+
+				// Use index instead of Add to handle duplicates
+				poms [final_artifact.ToString ()] = project;
+
+				logger.LogDebugMessage ("Registered POM for artifact '{0}' from '{1}'", final_artifact, filename);
+
+				return final_artifact;
+			}
+		} catch (Exception ex) {
+			logger.LogCodedError ("XA4246", Properties.Resources.XA4246, filename, ex.Message);
+			return null;
+		}
+	}
+
+	public Project ResolveRawProject (Artifact artifact)
+	{
+		if (poms.TryGetValue (artifact.ToString (), out var project))
+			return project;
+
+		logger.LogCodedError ("XA4247", Properties.Resources.XA4247, artifact);
+
+		throw new InvalidOperationException ($"No POM registered for {artifact}");
+	}
+}
+
+class MicrosoftNuGetPackageFinder
+{
+	readonly PackageListFile? package_list;
+
+	public MicrosoftNuGetPackageFinder (string? file, TaskLoggingHelper log)
+	{
+		if (file is null || !File.Exists (file)) {
+			log.LogMessage ("'microsoft-packages.json' file not found, Android NuGet suggestions will not be provided");
+			return;
+		}
+
+		try {
+			var json = File.ReadAllText (file);
+			package_list = JsonConvert.DeserializeObject<PackageListFile> (json);
+		} catch (Exception ex) {
+			log.LogMessage ("There was an error reading 'microsoft-packages.json', Android NuGet suggestions will not be provided: {0}", ex);
+		}
+	}
+
+	public string? GetNuGetPackage (string javaId)
+	{
+		return package_list?.Packages?.FirstOrDefault (p => p.JavaId?.Equals (javaId, StringComparison.OrdinalIgnoreCase) == true)?.NuGetId;
+	}
+
+	public class PackageListFile
+	{
+		[JsonProperty ("packages")]
+		public List<Package>? Packages { get; set; }
+	}
+
+	public class Package
+	{
+		[JsonProperty ("javaId")]
+		public string? JavaId { get; set; }
+
+		[JsonProperty ("nugetId")]
+		public string? NuGetId { get; set; }
+	}
+}
+
+public class NuGetPackageVersionFinder
+{
+	readonly LockFile lock_file;
+	readonly Dictionary<string, Artifact> cache = new Dictionary<string, Artifact> ();
+	readonly Regex tag = new Regex ("artifact_versioned=(?<GroupId>.+)?:(?<ArtifactId>.+?):(?<Version>.+)\\s?", RegexOptions.Compiled);
+	readonly Regex tag2 = new Regex ("artifact=(?<GroupId>.+)?:(?<ArtifactId>.+?):(?<Version>.+)\\s?", RegexOptions.Compiled);
```

Comment:
Can we use the source generator for new regexes:

https://learn.microsoft.com/en-us/dotnet/standard/base-types/regular-expression-source-generators

I don't know if we've tried this in an MSBuild task / netstandard2.0 project yet.

Reply:
It looks like `[GeneratedRegexAttribute]` is only available in .NET 7+:
https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.generatedregexattribute